### PR TITLE
Support binary boolean operators with nulls

### LINKED ISSUE #4241: Add support binary boolean operators with nulls
in datafsuion:
```
❯ select true or NULL ;
Plan("'Boolean OR Null' can't be evaluated because there isn't a common type to coerce the types to")
❯ select true or FALSE ;
+---------------------------------+
| Boolean(true) OR Boolean(false) |
+---------------------------------+
| true                            |
+---------------------------------+
1 row in set. Query took 0.003 seconds.
❯
```
RUN in spark-sql
```
spark-sql> select true or false;
true
Time taken: 3.139 seconds, Fetched 1 row(s)
spark-sql> select true or NULL;
true
Time taken: 0.076 seconds, Fetched 1 row(s)
spark-sql> select FALSE or NULL;
NULL
Time taken: 0.06 seconds, Fetched 1 row(s)
spark-sql>
```
I think its ok get error when run  `NULL or xxx`

_Originally posted by @Ted-Jiang in https://github.com/apache/arrow-datafusion/pull/4200#discussion_r1021374351_

### PR DESCRIPTION
**Pull request by Ted-Jiang**
Signed-off-by: yangjiang <yangjiang@ebay.com>

# Which issue does this PR close?

Closes #4241.

# Rationale for this change

in spark:
```
spark-sql> select null or true;
true
Time taken: 0.65 seconds, Fetched 1 row(s)
spark-sql> select null or false;
NULL
Time taken: 0.076 seconds, Fetched 1 row(s)
```
in pg:
```
postgres=# select true or NULL test;
 test
------
 t
(1 row)

postgres=# select false or NULL test;
 test
------

(1 row)
```
After this change in datafusion
```
❯ select NULL or false;
+------------------------+
| NULL OR Boolean(false) |
+------------------------+
|                        |
+------------------------+
1 row in set. Query took 0.067 seconds.
❯ select NULL or true;
+-----------------------+
| NULL OR Boolean(true) |
+-----------------------+
| true                  |
+-----------------------+
1 row in set. Query took 0.003 seconds.
```
Also test with `AND` 

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/type_coercion/binary.rs [hunk 1]
```diff
@@ -1022,6 +1026,30 @@ mod tests {
             Operator::Or,
             DataType::Boolean
         );
+        test_coercion_binary_rule!(
+            DataType::Boolean,
+            DataType::Null,
+            Operator::And,
+            DataType::Boolean
+        );
```
[the remark this answers is not in the document]
Comment (Ted-Jiang) [reply datafusion/expr/src/type_coercion/binary.rs:1034]: Thanks fix in [add ut](https://github.com/apache/arrow-datafusion/pull/4242/commits/58520a9edefe8a017cba02798d28dbb9af26c068)

**Review by andygrove — APPROVED** [review]

File: datafusion/expr/src/type_coercion/binary.rs [hunk 2]
```diff
@@ -98,8 +98,12 @@ pub fn coerce_types(
         | Operator::BitwiseShiftRight
         | Operator::BitwiseShiftLeft => bitwise_coercion(lhs_type, rhs_type),
         Operator::And | Operator::Or => match (lhs_type, rhs_type) {
-            // logical binary boolean operators can only be evaluated in bools
+            // logical binary boolean operators can only be evaluated in bools or nulls
             (DataType::Boolean, DataType::Boolean) => Some(DataType::Boolean),
+            (DataType::Null, DataType::Null) => Some(DataType::Null),
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/binary.rs:103]: 👍

**Review by alamb — COMMENTED** [review]
Thank you @Ted-Jiang

File: datafusion/expr/src/type_coercion/binary.rs [hunk 3]
```diff
@@ -98,8 +98,12 @@ pub fn coerce_types(
...
+            (DataType::Boolean, DataType::Null) | (DataType::Null, DataType::Boolean) => {
+                Some(DataType::Boolean)
```
Comment (jackwener) [inline datafusion/expr/src/type_coercion/binary.rs:105]: Here exist some problem.
true And Null -> Null
false Or Null -> Null

**Pull request merged**